### ISSUE TITLE
Use case / test results / future request re: custom_user

### ISSUE THREAD

**Issue by tc60045** [mbentley/docker-timemachine#66]
OK, I'm confused but not stuck -- I suppose.   

I've got a use case that doesn't quite fit the configuration here.
1.  I had a user "timemachine" on my local machine, part of group "backups"  UID=1105 / GID = 1105
2.  Every docker service I run with user level account UID = 1000 / GID = 1000
-->  I would prefer to re-use this existing user on my machine and not have this container create a new user. 
--> I'm not prepared to bind certain (unfamiliar) directories to make this happen.

Test attempts below.  TL;DR: OK to have username collide with existing username, provided you keep CUSTOM_USER as false and choose a GID that conforms to a GID for existing username, but not GID=UID for said username, which is unusual.

Future request:  change CUSTOM_USER to accept 4 cases:
- false:  just do your thing, but recall from testing below that UID 1000 / GID 1000 didn't work so something needs a tweak
- existing: user supplies an existing linux user account and the information is replicated in smb / avahi as appropriate
- bound: user supplies own linux account and binds things per your specifications
- group:  handle the situation where user wants lots of user accounts to be able to log in.  Just seems simpler than handling two binaries in the config, but, hey:  it is your pizza store (movie reference)

Thx!

---------- TEST RESULTS -----------

Case 1 - invent new user 'foobar' and try otherwise stock configuration:  CUSTOM_USER=false; TM_USERNAME=foobar; TM_GROUPNAME=foobar; password=password; UID = 1000; GID = 1000 

docker-compose logs look good, but it is impossible to log into the server: foobar / password doesn't work; foobar / '' doesn't work.  Also note the line for "not setting ownership and permissions for /opt/foobar" -- is that right?

docker-compose logs timemachine
timemachine           | INFO: CUSTOM_SMB_CONF=false; generating [global] section of /etc/samba/smb.conf...
timemachine           | Creating /var/log/samba/cores
timemachine           | INFO: Avahi - generating base configuration in /etc/avahi/services/smbd.service...
timemachine           | INFO: Avahi - adding the 'dk0', 'Veronica TM Server' share txt-record to /etc/avahi/services/smbd.service...
timemachine           | INFO: Group foobar doesn't exist; creating...
timemachine           | INFO: User foobar doesn't exist; creating...
timemachine           | INFO: Setting password from environment variable
timemachine           | chpasswd: password for 'foobar' changed
timemachine           | INFO: INFO: CUSTOM_SMB_CONF=false; generating [Veronica TM Server] section of /etc/samba/smb.conf...
timemachine           | INFO: Samba - Created User foobar password set to none.
timemachine           | INFO: Samba - Enabled user foobar.
timemachine           | INFO: Samba - setting password
timemachine           | INFO: SET_PERMISSIONS=false; not setting ownership and permissions for /opt/foobar
timemachine           | INFO: Avahi - completing the configuration in /etc/avahi/services/smbd.service...
timemachine           | INFO: entrypoint complete; executing 's6-svscan /etc/s6'
timemachine           | dbus socket not yet available; sleeping...
timemachine           | nmbd version 4.12.7 started.
timemachine           | Copyright Andrew Tridgell and the Samba Team 1992-2020
timemachine           | daemon_ready: daemon 'nmbd' finished starting up and ready to serve connections
timemachine           | smbd version 4.12.7 started.
timemachine           | Copyright Andrew Tridgell and the Samba Team 1992-2020
timemachine           | INFO: Profiling support unavailable in this build.
timemachine           | daemon_ready: daemon 'smbd' finished starting up and ready to serve connections
timemachine           | Failed to fetch record!
timemachine           | Found user 'avahi' (UID 86) and group 'avahi' (GID 86).
timemachine           | Successfully dropped root privileges.
timemachine           | avahi-daemon 0.8 starting up.
timemachine           | WARNING: No NSS support for mDNS detected, consider installing nss-mdns!
timemachine           | Loading service file /etc/avahi/services/smbd.service.
timemachine           | Joining mDNS multicast group on interface eth0.IPv4 with address 192.168.1.16.
timemachine           | New relevant interface eth0.IPv4 for mDNS.
timemachine           | Joining mDNS multicast group on interface lo.IPv4 with address 127.0.0.1.
timemachine           | New relevant interface lo.IPv4 for mDNS.
timemachine           | Network interface enumeration completed.
timemachine           | Registering new address record for 192.168.1.16 on eth0.IPv4.
timemachine           | Registering new address record for 127.0.0.1 on lo.IPv4.
timemachine           | Server startup complete. Host name is Veronica-TM.local. Local service cookie is 1294941489.
timemachine           | Service "Veronica-TM" (/etc/avahi/services/smbd.service) successfully established.

Case 2:  intentionally collide with existing user timemachine, group timemachine but without declaring a custom_user (keep as false); all else the same from above.

Seemingly works, but again cannot log in with known good linux username / password

Attaching to timemachine
timemachine           | INFO: CUSTOM_SMB_CONF=false; generating [global] section of /etc/samba/smb.conf...
timemachine           | Creating /var/log/samba/cores
timemachine           | INFO: Avahi - generating base configuration in /etc/avahi/services/smbd.service...
timemachine           | INFO: Avahi - adding the 'dk0', 'Veronica TM Server' share txt-record to /etc/avahi/services/smbd.service...
timemachine           | INFO: Group timemachine doesn't exist; creating...
timemachine           | INFO: User timemachine doesn't exist; creating...
timemachine           | INFO: Setting password from environment variable
timemachine           | chpasswd: password for 'timemachine' changed
timemachine           | INFO: INFO: CUSTOM_SMB_CONF=false; generating [Veronica TM Server] section of /etc/samba/smb.conf...
timemachine           | INFO: Samba - Created User timemachine password set to none.
timemachine           | INFO: Samba - Enabled user timemachine.
timemachine           | INFO: Samba - setting password
timemachine           | INFO: SET_PERMISSIONS=false; not setting ownership and permissions for /opt/timemachine
timemachine           | INFO: Avahi - completing the configuration in /etc/avahi/services/smbd.service...
timemachine           | INFO: entrypoint complete; executing 's6-svscan /etc/s6'
timemachine           | dbus socket not yet available; sleeping...
timemachine           | nmbd version 4.12.7 started.
timemachine           | Copyright Andrew Tridgell and the Samba Team 1992-2020
timemachine           | daemon_ready: daemon 'nmbd' finished starting up and ready to serve connections
timemachine           | smbd version 4.12.7 started.
timemachine           | Copyright Andrew Tridgell and the Samba Team 1992-2020
timemachine           | INFO: Profiling support unavailable in this build.
timemachine           | daemon_ready: daemon 'smbd' finished starting up and ready to serve connections
timemachine           | Failed to fetch record!
timemachine           | Found user 'avahi' (UID 86) and group 'avahi' (GID 86).
timemachine           | Successfully dropped root privileges.
timemachine           | avahi-daemon 0.8 starting up.
timemachine           | WARNING: No NSS support for mDNS detected, consider installing nss-mdns!
timemachine           | Loading service file /etc/avahi/services/smbd.service.
timemachine           | Joining mDNS multicast group on interface eth0.IPv4 with address 192.168.1.16.
timemachine           | New relevant interface eth0.IPv4 for mDNS.
timemachine           | Joining mDNS multicast group on interface lo.IPv4 with address 127.0.0.1.
timemachine           | New relevant interface lo.IPv4 for mDNS.
timemachine           | Network interface enumeration completed.
timemachine           | Registering new address record for 192.168.1.16 on eth0.IPv4.
timemachine           | Registering new address record for 127.0.0.1 on lo.IPv4.
timemachine           | Server startup complete. Host name is Veronica-TM.local. Local service cookie is 3803798048.
timemachine           | Service "Veronica-TM" (/etc/avahi/services/smbd.service) successfully established.

Case 3:  intentional collision (username = existing linux username 'timemachine') but now set UID and GID to 1105, which are the right UID and GID for id timemachine, sez my linux box.

no dice -- again cannot log in.  Output the same

Case 4:  intentional collision (username = existing linux username 'timemachine') but now set UID and GID to 1100, which is an odd UID (there is no UID 1100 on my system), but it is the GID for (existing) group backups.

Worked.   I can now log in.  I lucked into this setup last night and had to test my way back to it today.  
BUT -- linux owner is now '1100' ?!

timemachine           | INFO: CUSTOM_SMB_CONF=false; generating [global] section of /etc/samba/smb.conf...
timemachine           | Creating /var/log/samba/cores
timemachine           | INFO: Avahi - generating base configuration in /etc/avahi/services/smbd.service...
timemachine           | INFO: Avahi - adding the 'dk0', 'Veronica TM Server' share txt-record to /etc/avahi/services/smbd.service...
timemachine           | INFO: Group timemachine doesn't exist; creating...
timemachine           | INFO: User timemachine doesn't exist; creating...
timemachine           | INFO: Setting password from environment variable
timemachine           | chpasswd: password for 'timemachine' changed
timemachine           | INFO: INFO: CUSTOM_SMB_CONF=false; generating [Veronica TM Server] section of /etc/samba/smb.conf...
timemachine           | INFO: Samba - Created User timemachine password set to none.
timemachine           | INFO: Samba - Enabled user timemachine.
timemachine           | INFO: Samba - setting password
timemachine           | INFO: SET_PERMISSIONS=false; not setting ownership and permissions for /opt/timemachine
timemachine           | INFO: Avahi - completing the configuration in /etc/avahi/services/smbd.service...
timemachine           | INFO: smbd PID exists; removing...
timemachine           | removed '/run/samba/smbd.pid'
timemachine           | INFO: entrypoint complete; executing 's6-svscan /etc/s6'
timemachine           | dbus socket not yet available; sleeping...
timemachine           | nmbd version 4.12.7 started.
timemachine           | Copyright Andrew Tridgell and the Samba Team 1992-2020
timemachine           | daemon_ready: daemon 'nmbd' finished starting up and ready to serve connections
timemachine           | smbd version 4.12.7 started.
timemachine           | Copyright Andrew Tridgell and the Samba Team 1992-2020
timemachine           | INFO: Profiling support unavailable in this build.
timemachine           | daemon_ready: daemon 'smbd' finished starting up and ready to serve connections
timemachine           | Failed to fetch record!
timemachine           | Found user 'avahi' (UID 86) and group 'avahi' (GID 86).
timemachine           | Successfully dropped root privileges.
timemachine           | avahi-daemon 0.8 starting up.
timemachine           | WARNING: No NSS support for mDNS detected, consider installing nss-mdns!
timemachine           | Loading service file /etc/avahi/services/smbd.service.
timemachine           | Joining mDNS multicast group on interface eth0.IPv4 with address 192.168.1.16.
timemachine           | New relevant interface eth0.IPv4 for mDNS.
timemachine           | Joining mDNS multicast group on interface lo.IPv4 with address 127.0.0.1.
timemachine           | New relevant interface lo.IPv4 for mDNS.
timemachine           | Network interface enumeration completed.
timemachine           | Registering new address record for 192.168.1.16 on eth0.IPv4.
timemachine           | Registering new address record for 127.0.0.1 on lo.IPv4.
timemachine           | Server startup complete. Host name is Veronica-TM.local. Local service cookie is 2166607592.
timemachine           | Service "Veronica-TM" (/etc/avahi/services/smbd.service) successfully established.
timemachine           | Failed to fetch record!
timemachine           | pcap cache not loaded
timemachine           | *****
timemachine           | 
timemachine           | Samba name server VERONICA-TM is now a local master browser for workgroup WORKGROUP on subnet 192.168.1.16
timemachine           | 
timemachine           | *****

case 5: intentional collision; keep UID / GUID the same as previous case; try CUSTOM_USER = true

Cannot connect.  Fails to find Unix account for timemachine! 

timemachine           | INFO: CUSTOM_SMB_CONF=false; generating [global] section of /etc/samba/smb.conf...
timemachine           | Creating /var/log/samba/cores
timemachine           | INFO: Avahi - generating base configuration in /etc/avahi/services/smbd.service...
timemachine           | INFO: Avahi - adding the 'dk0', 'Veronica TM Server' share txt-record to /etc/avahi/services/smbd.service...
timemachine           | INFO: CUSTOM_USER=true; skipping user, group, and data directory creation; using pre-existing values in /etc/passwd, /etc/group, and /etc/shadow
timemachine           | INFO: CUSTOM_SMB_CONF=false; generating [Veronica TM Server] section of /etc/samba/smb.conf...
timemachine           | Failed to find a Unix account for timemachine
timemachine           | Failed to find a Unix account for timemachine
timemachine           | INFO: Samba - Created User timemachine password set to none.
timemachine           | Failed to find a Unix account for timemachine
timemachine           | Failed to find a Unix account for timemachine
timemachine           | Failed to find a Unix account for timemachine
timemachine           | INFO: Samba - Enabled user timemachine.
timemachine           | INFO: Samba - setting password
timemachine           | Failed to find a Unix account for timemachine
timemachine           | Failed to find a Unix account for timemachine
timemachine           | INFO: SET_PERMISSIONS=false; not setting ownership and permissions for /opt/timemachine
timemachine           | INFO: Avahi - completing the configuration in /etc/avahi/services/smbd.service...
timemachine           | INFO: smbd PID exists; removing...
timemachine           | removed '/run/samba/smbd.pid'
timemachine           | INFO: entrypoint complete; executing 's6-svscan /etc/s6'
timemachine           | dbus socket not yet available; sleeping...
timemachine           | nmbd version 4.12.7 started.
timemachine           | Copyright Andrew Tridgell and the Samba Team 1992-2020
timemachine           | daemon_ready: daemon 'nmbd' finished starting up and ready to serve connections
timemachine           | smbd version 4.12.7 started.
timemachine           | Copyright Andrew Tridgell and the Samba Team 1992-2020
timemachine           | INFO: Profiling support unavailable in this build.
timemachine           | daemon_ready: daemon 'smbd' finished starting up and ready to serve connections
timemachine           | Failed to fetch record!
timemachine           | Found user 'avahi' (UID 86) and group 'avahi' (GID 86).
timemachine           | Successfully dropped root privileges.
timemachine           | avahi-daemon 0.8 starting up.
timemachine           | WARNING: No NSS support for mDNS detected, consider installing nss-mdns!
timemachine           | Loading service file /etc/avahi/services/smbd.service.
timemachine           | Joining mDNS multicast group on interface eth0.IPv4 with address 192.168.1.16.
timemachine           | New relevant interface eth0.IPv4 for mDNS.
timemachine           | Joining mDNS multicast group on interface lo.IPv4 with address 127.0.0.1.
timemachine           | New relevant interface lo.IPv4 for mDNS.
timemachine           | Network interface enumeration completed.
timemachine           | Registering new address record for 192.168.1.16 on eth0.IPv4.
timemachine           | Registering new address record for 127.0.0.1 on lo.IPv4.
timemachine           | Server startup complete. Host name is Veronica-TM.local. Local service cookie is 3836076221.
timemachine           | Service "Veronica-TM" (/etc/avahi/services/smbd.service) successfully established.

**Comment by mbentley** [comment]
tl;dr - There is no connection between usernames and passwords from the host and the container unless you set the usernames and passwords the same on the host and inside the container.  There is a connection between the UID and GID values of persistent files stored on the host's disk.  Longer explanation at the end.

I'm going to reply to each one of these at a time.  I'm going to use `docker run...` commands because it is easier for me.  "not setting ownership and permissions for /opt/foobar" is correct because the default setting is to not have it automatically set the permissions.  Set "SET_PERMISSIONS" to true if you want that.

Case 1 - I can login just fine with my defined username and password in the env vars.

```
docker run -d --restart=always \
  --name timemachine \
  --net=host \
  -e CUSTOM_SMB_CONF="false" \
  -e CUSTOM_USER="false" \
  -e DEBUG_LEVEL="1" \
  -e MIMIC_MODEL="TimeCapsule8,119" \
  -e EXTERNAL_CONF="" \
  -e HIDE_SHARES="no" \
  -e TM_USERNAME="foobar" \
  -e TM_GROUPNAME="foobar" \
  -e TM_UID="1000" \
  -e TM_GID="1000" \
  -e PASSWORD="password" \
  -e SET_PERMISSIONS="false" \
  -e SHARE_NAME="TimeMachine" \
  -e SMB_PORT="445" \
  -e VOLUME_SIZE_LIMIT="0" \
  -e WORKGROUP="WORKGROUP" \
  -v timemachine:/opt/foobar \
  -v timemachine-var-lib-samba:/var/lib/samba \
  -v timemachine-var-cache-samba:/var/cache/samba \
  -v timemachine-run-samba:/run/samba \
  mbentley/timemachine:smb
```

Case 2 - For me, this is the same as case 1 because `1000:1000` already maps to my default login username on the host.  I can login just fine with my defined username and password in the env vars.  I am not using the username and password of that user that is defined on the host because that isn't how it works.

Cases 2, 3, and 4 - there is not really any user collision per se.  Users that authenticate to the container have their own usernames.  The only thing that "collides" is that the persistent file storage will have the same UID/GID.  Unless you're mounting /etc/passwd and /etc/group from the host, then you're not going to have actual username and password collisions.

It sounds like you're not sure of what happens in Docker.  If I create a user that is `foobar` in a container with UID/GID of `1000:1000`, then all of that user's files will have the UID/GID of 1000:1000.  If I have a user named `user` on my host with the same UID/GID, the OS will map the file ownership as it is stored on the attributes of the files (`1000:1000`) to what is in /etc/passwd and /etc/group.  If you look in the container, it will show it is owned as `foobar` and from the host, it will show as `user`.  Use `ls -lan` and it will list the numerical UID/GID values of the files in both scenarios.  They are the same because they're just UIDs and GIDs.  They're just getting mapped to user friendly values of the user name and group name that is defined in /etc/passwd and /etc/group.

For Case 5, you need to use the `EXTERNAL_CONF` where you pass your own users in files as described [here](https://github.com/mbentley/docker-timemachine#adding-multiple-users--shares)  You have to tell it what users to have it generate.  If it generates no users, you'll have none.

**Comment by tc60045** [comment]
Matt, thanks for your response.  A few things:
- my thoughts on case 1 were along the lines of "what would a script kiddie do when encountering Matt's work?"  Slap in some new usernames / passwords and expect it all to work.  So take that as note to future Matt:  someone is going to assume and ask..

- cases 2 - 4 are interesting, to me, as I would expect a user to walk in with an existing linux user.  The UID / GID part needs some investigation (yes, I do know how they work... I pass UID/GID=1000 along with no worry to two dozen containers that I've used for a decade).
On my machine:
backups group = GID 1100 / no user id (Ubuntu standard)
timemachine = UID 1105 / GID 1105
passing 1105 / 1105 should have worked, but I had to pass 1100 / 1100 to make it work (and again, there is no UID = 1100 on a stock Ubuntu system).  

Weirdly,  thereafter, files that are written over the connection appear with a username of 1100 (in case 4).   I can't (yet) figure out where in your code to find this is happening -- is there a loop that is chown'ing files in the mount with that UID?  I don't see that.  

![by default 2020-11-11 at 1 24 05 PM](https://user-images.githubusercontent.com/2856318/98885801-4d727200-2458-11eb-8afb-a988385aa428.jpg)

If not, I would have expected that when I logged into shell in the container, there would be a user '1100'.  Nope.  I see:

$id
uid=0(root) gid=0(root) groups=0(root),1(bin),2(daemon),3(sys),4(adm),6(disk),10(wheel),11(floppy),20(dialout),26(tape),27(video)

How is this happening?  I don't know -- and I discovered the limit of my docker container user permission knowledge when I tested:  touch /opt/$username foo.blah.   The username correspondingly came up in ls as root and that persisted when I exited and ls'd the mount -- had to sudo to change it back.  Welcome to userspace in docker?  

Everything is workable at the moment, so I'm happy to move along.  But I'm also convinced that something is lurking both on setup with requiring different UID / GID than those that = a linux account and with the strange username on files created through the connection.   All the best!

**Comment by mbentley** [comment]
I am still not following the connection on an existing Linux user exactly but it appears that you're specifically wanting to use existing UIDs and the backup GID in the other use cases that you present.

I am not sure why it would fail with any specific UID.  By default, the alpine linux image doesn't have any users or groups over 999 in use out of the box so there shouldn't be an existing conflicts.

In the case of the files being written as 1100, that is going to be happening because that is the UID of the user that was created inside the container.  When you write to a SMB share, it will write files and directories as the user that is connected to the share.

When you get failed login attempts, are you seeing any logs like this?

```
add_local_groups: SID S-1-5-21-65961667-3532078895-128230851-1000 -> getpwuid(1001) failed, is nsswitch configured?
add_local_groups: SID S-1-5-21-65961667-3532078895-128230851-1000 -> getpwuid(1001) failed, is nsswitch configured?
```

If so, it is probably because of cached data in the `/var/cache/samba` volume between changing settings.  That can prevent you from logging in.  Make sure to prune the non-data volumes when changing settings.

And when you exec into the container, it is running as root because that is the default user that the entrypoint runs as.  You have to be root to create the necessary processes and then when users connect, files are written as them.  Samba doesn't run specifically as a user's UID in order to function.

I was just able to test a multi-user example:

```
docker run -d --restart=always \
  --name timemachine \
  --net=host \
  -e CUSTOM_SMB_CONF="false" \
  -e CUSTOM_USER="false" \
  -e DEBUG_LEVEL="1" \
  -e MIMIC_MODEL="TimeCapsule8,119" \
  -e EXTERNAL_CONF="/multi-user" \
  -e HIDE_SHARES="no" \
  -e SET_PERMISSIONS="false" \
  -e SHARE_NAME="TimeMachine" \
  -e SMB_PORT="445" \
  -e VOLUME_SIZE_LIMIT="0" \
  -e WORKGROUP="WORKGROUP" \
  -v timemachine:/opt \
  -v timemachine-var-lib-samba:/var/lib/samba \
  -v timemachine-var-cache-samba:/var/cache/samba \
  -v timemachine-run-samba:/run/samba \
  -v /zfs/storage/tm/multi-user:/multi-user:ro \
  mbentley/timemachine:smb
```

Where my multi-user files look like:

`user1.conf`
```
TM_USERNAME=foo1
TM_GROUPNAME=foo1group
PASSWORD=foopass
SHARE_NAME=foo1
VOLUME_SIZE_LIMIT="1 T"
TM_UID=1001
TM_GID=1001
```

`user2.conf`
```
TM_USERNAME=foo2
TM_GROUPNAME=foo2group
PASSWORD=foopass
SHARE_NAME=foo2
VOLUME_SIZE_LIMIT="1 T"
TM_UID=1002
TM_GID=1002
```

`user3.conf`
```
TM_USERNAME=foo3
TM_GROUPNAME=foo3group
PASSWORD=foopass
SHARE_NAME=foo3
VOLUME_SIZE_LIMIT="1 T"
TM_UID=1003
TM_GID=1003
```

`user4.conf`
```
TM_USERNAME=foo4
TM_GROUPNAME=foo4group
PASSWORD=foopass
SHARE_NAME=foo4
VOLUME_SIZE_LIMIT="1 T"
TM_UID=1004
TM_GID=1004
```

`user5.conf`
```
TM_USERNAME=foo5
TM_GROUPNAME=foo5group
PASSWORD=foopass
SHARE_NAME=foo5
VOLUME_SIZE_LIMIT="1 T"
TM_UID=1005
TM_GID=1005
```

And I wrote files as two different users: `foo1` and `foo2` to their respective shares, looking at the files from the host OS itself:
```
root@dockerlab [ssh] /v/l/d/v/t/_data# ll
total 24
drwxr-sr-x 2 1001 1001 4096 Nov 12 06:26 foo1/
drwxr-sr-x 2 1002 1002 4096 Nov 12 06:29 foo2/
drwxr-sr-x 2 1003 1003 4096 Nov 12 06:23 foo3/
drwxr-sr-x 2 1004 1004 4096 Nov 12 06:23 foo4/
drwxr-sr-x 2 1005 1005 4096 Nov 12 06:23 foo5/
drwxr-xr-x 2 root root 4096 Oct 24 01:03 timemachine/
root@dockerlab [ssh] /v/l/d/v/t/_data# ll foo1
total 8
-rw-r--r--+ 1 1001 1001 1150 Nov 29  2076 favicon.ico
root@dockerlab [ssh] /v/l/d/v/t/_data# ll foo2/
total 8
-rw-r--r--+ 1 1002 1002 1150 Nov 29  2076 favicon.ico
```

As you can see, I have no users on the host that map to those users.

That being said, I did find a bug there where it is creating a `timemachine` directory when it shouldn't be.

**Comment by tc60045** [comment]
Matt, thanks -- I'm grateful for the example, too, as it shows what I am seeing:  the username (not ID) that is associated with those files does not conform to the host users.  In my case, 1100 is the user, and that user does not exist on host; in your example, '1101' is the "user" associated with foo1/favicon.ico, not 'foo1'. 

Oddly, when I exec into the docker image (I realize this is always as root :P ), and type 'id', I do not see the user 1100 and I don't see any line of code that deletes a user account.  When I 'cat /etc/group' I see a group named timemachine with the GID of 1100.  
(base) todd@veronica:~/scripts/docker$ docker-compose exec timemachine sh
/ # cat /etc/group
root:x:0:root
bin:x:1:root,bin,daemon
daemon:x:2:root,bin,daemon
sys:x:3:root,bin,adm
adm:x:4:root,adm,daemon
tty:x:5:
disk:x:6:root,adm
lp:x:7:lp
mem:x:8:
kmem:x:9:
wheel:x:10:root
floppy:x:11:root
mail:x:12:mail
news:x:13:news
uucp:x:14:uucp
man:x:15:man
cron:x:16:cron
console:x:17:
audio:x:18:
cdrom:x:19:
dialout:x:20:root
ftp:x:21:
sshd:x:22:
input:x:23:
at:x:25:at
tape:x:26:root
video:x:27:root
netdev:x:28:
readproc:x:30:
squid:x:31:squid
xfs:x:33:xfs
kvm:x:34:kvm
games:x:35:
shadow:x:42:
cdrw:x:80:
usb:x:85:
vpopmail:x:89:
users:x:100:games
ntp:x:123:
nofiles:x:200:
smmsp:x:209:smmsp
locate:x:245:
abuild:x:300:
utmp:x:406:
ping:x:999:
nogroup:x:65533:
nobody:x:65534:
avahi:x:86:avahi
messagebus:x:101:messagebus
catchlog:x:102:catchlog
timemachine:x:1100:timemachine
/ # id
uid=0(root) gid=0(root) groups=0(root),1(bin),2(daemon),3(sys),4(adm),6(disk),10(wheel),11(floppy),20(dialout),26(tape),27(video)
/ # 

Can you post the same for your example with the foos?  Thanks!

**Comment by tc60045** [comment]
<edited by mbentley> on the above = thank you for saving me from this vile interface!

**Comment by mbentley** [comment]
Right, you should not see the user `1100` in the output of `id` as it is going to show the UID and groups/GIDs that the current user (root) is a member of; nothing to do with the user that maps to `1100`.

> in your example, '1101' is the "user" associated with foo1/favicon.ico, not 'foo1'.

Right.  `foo1` doesn't exist on my host; just in the container.

```
/ # cat /etc/passwd
root:x:0:0:root:/root:/bin/ash
bin:x:1:1:bin:/bin:/sbin/nologin
daemon:x:2:2:daemon:/sbin:/sbin/nologin
adm:x:3:4:adm:/var/adm:/sbin/nologin
lp:x:4:7:lp:/var/spool/lpd:/sbin/nologin
sync:x:5:0:sync:/sbin:/bin/sync
shutdown:x:6:0:shutdown:/sbin:/sbin/shutdown
halt:x:7:0:halt:/sbin:/sbin/halt
mail:x:8:12:mail:/var/mail:/sbin/nologin
news:x:9:13:news:/usr/lib/news:/sbin/nologin
uucp:x:10:14:uucp:/var/spool/uucppublic:/sbin/nologin
operator:x:11:0:operator:/root:/sbin/nologin
man:x:13:15:man:/usr/man:/sbin/nologin
postmaster:x:14:12:postmaster:/var/mail:/sbin/nologin
cron:x:16:16:cron:/var/spool/cron:/sbin/nologin
ftp:x:21:21::/var/lib/ftp:/sbin/nologin
sshd:x:22:22:sshd:/dev/null:/sbin/nologin
at:x:25:25:at:/var/spool/cron/atjobs:/sbin/nologin
squid:x:31:31:Squid:/var/cache/squid:/sbin/nologin
xfs:x:33:33:X Font Server:/etc/X11/fs:/sbin/nologin
games:x:35:35:games:/usr/games:/sbin/nologin
cyrus:x:85:12::/usr/cyrus:/sbin/nologin
vpopmail:x:89:89::/var/vpopmail:/sbin/nologin
ntp:x:123:123:NTP:/var/empty:/sbin/nologin
smmsp:x:209:209:smmsp:/var/spool/mqueue:/sbin/nologin
guest:x:405:100:guest:/dev/null:/sbin/nologin
nobody:x:65534:65534:nobody:/:/sbin/nologin
avahi:x:86:86:Avahi System User:/var/run/avahi-daemon:/sbin/nologin
messagebus:x:100:101:messagebus:/dev/null:/sbin/nologin
catchlog:x:101:102:catchlog:/home/catchlog:/bin/false
foo1:x:1001:1001:Linux User,,,:/opt/foo1:/bin/false
foo2:x:1002:1002:Linux User,,,:/opt/foo2:/bin/false
foo3:x:1003:1003:Linux User,,,:/opt/foo3:/bin/false
foo4:x:1004:1004:Linux User,,,:/opt/foo4:/bin/false
foo5:x:1005:1005:Linux User,,,:/opt/foo5:/bin/false
```

```
/ # cat /etc/group
root:x:0:root
bin:x:1:root,bin,daemon
daemon:x:2:root,bin,daemon
sys:x:3:root,bin,adm
adm:x:4:root,adm,daemon
tty:x:5:
disk:x:6:root,adm
lp:x:7:lp
mem:x:8:
kmem:x:9:
wheel:x:10:root
floppy:x:11:root
mail:x:12:mail
news:x:13:news
uucp:x:14:uucp
man:x:15:man
cron:x:16:cron
console:x:17:
audio:x:18:
cdrom:x:19:
dialout:x:20:root
ftp:x:21:
sshd:x:22:
input:x:23:
at:x:25:at
tape:x:26:root
video:x:27:root
netdev:x:28:
readproc:x:30:
squid:x:31:squid
xfs:x:33:xfs
kvm:x:34:kvm
games:x:35:
shadow:x:42:
cdrw:x:80:
usb:x:85:
vpopmail:x:89:
users:x:100:games
ntp:x:123:
nofiles:x:200:
smmsp:x:209:smmsp
locate:x:245:
abuild:x:300:
utmp:x:406:
ping:x:999:
nogroup:x:65533:
nobody:x:65534:
avahi:x:86:avahi
messagebus:x:101:messagebus
catchlog:x:102:catchlog
foo1group:x:1001:foo1
foo2group:x:1002:foo2
foo3group:x:1003:foo3
foo4group:x:1004:foo4
foo5group:x:1005:foo5
```

**Comment by mbentley** [comment]
I think all of the use cases are taken care of, would you agree or am I missing something?

**Comment by MSSUnQ** [comment]
Clear Winbind cache
How to clear Winbind cache
 

Stop the Winbind service :

[root@danny root]#  service winbind stop
Stop the samba service if you have it :

[root@danny root]#  service smb stop
Clear the Samba Net cache:

[root@danny root]#  net cache flush
Delete the Winbind cache :

[root@danny samba]# cd /var/lib/samba/
Always take backup before deleting anything:

[root@danny samba]# mkdir old && mv *.tdb old/.
Start the samba service if you have it :

[root@danny samba]# service smb start
Start the Winbind service

[root@danny samba]# service winbind start
Now you can check.
